File: algorithm/study/Core-05-Simulation/14503/python/src/solution.py
import sys
input = sys.stdin.readline

def main():
    N, M = map(int, input().split())
    r, c, d = map(int, input().split())
    grid = [list(map(int, input().split())) for _ in range(N)]

    # 방향 인코딩: 0=N, 1=E, 2=S, 3=W
    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]

    cleaned = [[False] * M for _ in range(N)]
    count = 0

    while True:
        # 1단계: 현재 칸을 청소
        if not cleaned[r][c]:
            cleaned[r][c] = True
            count += 1

        # 2단계: 왼쪽 회전을 최대 4번 시도
        found = False
        for _ in range(4):
            d = (d + 3) % 4  # 왼쪽으로 회전
            nr, nc = r + dr[d], c + dc[d]
            if 0 <= nr < N and 0 <= nc < M and grid[nr][nc] == 0 and not cleaned[nr][nc]:
                r, c = nr, nc
                found = True
                break

        if found:
            continue

        # 3단계: 네 방향이 모두 막히면 뒤로 이동을 시도
        bd = (d + 2) % 4  # 반대 방향
        br, bc = r + dr[bd], c + dc[bd]
        if 0 <= br < N and 0 <= bc < M and grid[br][bc] != 1:
            r, c = br, bc
        else:
            break  # 뒤가 벽이면 종료

    print(count)

if __name__ == "__main__":
    main()

File: algorithm/study/Core-05-Simulation/14503/cpp/src/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M; cin >> N >> M;
    int r, c, d; cin >> r >> c >> d;

    vector<vector<int>> grid(N, vector<int>(M));
    for(int i = 0; i < N; i++)
        for(int j = 0; j < M; j++)
            cin >> grid[i][j];

    int dr[] = {-1, 0, 1, 0};
    int dc[] = {0, 1, 0, -1};
    vector<vector<bool>> cleaned(N, vector<bool>(M, false));
    int count = 0;

    while(true){
        // 현재 칸을 청소
        if(!cleaned[r][c]){
            cleaned[r][c] = true;
            count++;
        }

        // 왼쪽으로 4번 회전해 본다
        bool found = false;
        for(int i = 0; i < 4; i++){
            d = (d + 3) % 4;
            int nr = r + dr[d], nc = c + dc[d];
            if(nr >= 0 && nr < N && nc >= 0 && nc < M
               && grid[nr][nc] == 0 && !cleaned[nr][nc]){
                r = nr; c = nc;
                found = true;
                break;
            }
        }
        if(found) continue;

        // 네 방향이 모두 막히면 뒤로 한 칸 시도한다.
        int bd = (d + 2) % 4;
        int br = r + dr[bd], bc = c + dc[bd];
        if(br >= 0 && br < N && bc >= 0 && bc < M && grid[br][bc] != 1){
            r = br; c = bc;
        } else {
            break;
        }
    }

    cout << count << "\n";
    return 0;
}
